runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1651,6 +1707,277 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+    * n3042n          ==>aladd
+    */
+   TR::Node *indexNode = callNode->getChild(0);
+   TR::Node *arrayRefNode = callNode->getChild(1);
+
+   TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayClass);
+
+   const TR::TypeLayout *fieldTypeLayout = comp()->typeLayout(arrayComponentClass);
+   size_t fieldCount = fieldTypeLayout->count();
+
+   TR::ResolvedMethodSymbol *method = comp()->getOwningMethodSymbol(callNode->getOwningMethod());
+
+   // create loadaddr
+   TR::Node *classNode = TR::Node::createWithSymRef(callNode, TR::loadaddr, 0,
+                                            comp()->getSymRefTab()->findOrCreateClassSymbol(method, -1, arrayComponentClass));
+
+   // Create newvalue
+   TR::Node *newValueNode = TR::Node::recreateWithoutProperties(callNode, TR::newvalue, fieldCount+1, classNode, comp()->getSymRefTab()->findOrCreateNewValueSymbolRef(method));
+   newValueNode->setIdentityless(true);
+
+   // If the array element contains zero field, the newvalue will contain only loadaddr
+   if (fieldCount == 0)
+      {
+      if (trace())
+         {
+         traceMsg(comp(),"%s fieldCount 0: The call node is recreated to newValueNode n%dn\n", __FUNCTION__, newValueNode->getGlobalIndex());
+         }
+      return;
+      }
+
+   // Generate aladd for element address
+   int32_t elementStride = TR::Compiler->cls.flattenedArrayElementSize(comp(), arrayClass);
+   TR::Node *elementAddressNode = J9::TransformUtil::calculateElementAddressWithElementStride(comp(), arrayRefNode, indexNode, elementStride);
+
+   // Generate load for each of the final fields
+   int newValueNodeChildIndex = 1;
+   for (size_t idx = 0; idx < fieldCount; idx++)
+      {
+      const TR::TypeLayoutEntry &fieldEntry = fieldTypeLayout->entry(idx);
+      TR::ILOpCodes loadOpCode = comp()->il.opCodeForIndirectLoad(fieldEntry._datatype);
+
+      auto * fieldSymRef = comp()->getSymRefTab()->findOrFabricateFlattenedArrayElementFieldShadowSymbol(arrayComponentClass,
+                                                                  fieldEntry._datatype,
+                                                                  fieldEntry._offset,
+                                                                  fieldEntry._isPrivate,
+                                                                  fieldEntry._fieldname,
+                                                                  fieldEntry._typeSignature);
+      if (trace())
+         {
+         traceMsg(comp(), "%s %s fieldSymRef: %s fieldEntry[%d] fieldName %s fieldSig %s type %d offset %d\n", __FUNCTION__,
+            comp()->getDebug()->getName(loadOpCode), comp()->getDebug()->getName(fieldSymRef), (int)idx,
+            fieldEntry._fieldname, fieldEntry._typeSignature, fieldEntry._datatype.getDataType(), fieldEntry._offset);
+         }
+
+      // There's no need to apply a NULLCHK on the load because the value is known to be a value type, which is never null.
```

Comment:
Sorry - the suggestion to add a new comment regarding the lack of a NULLCHK was meant to apply to the value being stored in `transformFlattenedArrayElementStore`.  I'll add a comment below.

Reply:
Resolved in bdab531